lib/mayaUsd/ufe/UsdUndoRenameCommand.cpp

```diff
@@ -42,44 +42,49 @@ namespace ufe {
 
 UsdUndoRenameCommand::UsdUndoRenameCommand(const UsdSceneItem::Ptr& srcItem, const Ufe::PathComponent& newName)
     : Ufe::UndoableCommand()
+    , _ufeSrcItem(srcItem)
+    , _ufeDstItem(nullptr)
+    , _prim(srcItem->prim())
+    , _stage(_prim.GetStage())
+    , _newName(newName.string())
+    , _oldName(_prim.GetName().GetString())
 {
-    const UsdPrim& prim = srcItem->prim();
-    _stage = prim.GetStage();
-    _ufeSrcItem = srcItem;
-    _usdSrcPath = prim.GetPath();
-
-    // Every call to rename() (through execute(), undo() or redo()) removes
-    // a prim, which becomes expired.  Since USD UFE scene items contain a
-    // prim, we must recreate them after every call to rename.
-    _usdDstPath = prim.GetParent().GetPath().AppendChild(TfToken(newName.string()));
-
-    _layer = MayaUsdUtils::defPrimSpecLayer(prim);
-    if (!_layer) {
-        std::string err = TfStringPrintf("No prim found at %s", prim.GetPath().GetString().c_str());
-        throw std::runtime_error(err.c_str());
-    }
-
     // if the current layer doesn't have any opinions that affects selected prim
-    if (!MayaUsdUtils::doesEditTargetLayerHavePrimSpec(prim)) {
-        auto possibleTargetLayer = MayaUsdUtils::strongestLayerWithPrimSpec(prim);
+    if (!MayaUsdUtils::doesEditTargetLayerHavePrimSpec(_prim)) {
+        auto possibleTargetLayer = MayaUsdUtils::strongestLayerWithPrimSpec(_prim);
         std::string err = TfStringPrintf("Cannot rename [%s] defined on another layer. " 
                                          "Please set [%s] as the target layer to proceed", 
-                                         prim.GetName().GetString().c_str(),
+                                         _prim.GetName().GetString().c_str(),
                                          possibleTargetLayer->GetDisplayName().c_str());
         throw std::runtime_error(err.c_str());
     }
     else
     {
-        auto layers = MayaUsdUtils::layersWithPrimSpec(prim);
+        // account for internal vs external references
+        if (_prim.HasAuthoredReferences()) {
+            auto primSpec = MayaUsdUtils::getPrimSpecAtEditTarget(_stage, _prim);
```

Comment:
Given that you've submitted a pull request to Pixar to better handle this functionality, wouldn't it be good to encapsulate this behind a utility function so that this code won't have to change when internal reference testing improves?  The encapsulation would be specially useful since you'll have to support different implementations, depending on the version of USD that's being used.  E.g. something like
bool isInternalReference(primSpec)

Reply:
Good point, will make one in the next commit.

Reply:
@ppt-adsk please see a305af95bbe17a8e7a68e8e807335ccc80fe043f